app/models/manageiq/providers/openstack/cloud_manager/vm.rb

```diff
@@ -36,6 +36,55 @@ def floating_ip
     floating_ips.first
   end
 
+  def associate_floating_ip_from_network(public_network, port = nil)
+    ext_management_system.with_provider_connection(:service     => "Network",
+                                                   :tenant_name => cloud_tenant.name) do |connection|
+      unless port
+        network_ports.each do |network_port|
+          # Cycle through all ports and find one that is actually connected to the public network with router,
+          if network_port.public_networks.detect { |x| x.try(:ems_ref) == public_network.ems_ref }
+            port = network_port
+            break
+          end
+        end
+      end
+      unless port
+        raise(MiqException::MiqNetworkPortNotDefinedError,
+              "Neutron port for floating IP association is not defined for OpenStack"\
+              "network #{public_network.ems_ref} and EMS '#{ext_management_system.name}'")
+      end
+
+      connection.create_floating_ip(public_network.ems_ref, :port_id => port.ems_ref)
+    end
+  end
+
+  def delete_floating_ips(floating_ips)
+    # TODO(lsmola) we have the method here because we need to take actual cloud_tenant from the VM.
+    # This should be refactored to FloatingIP, when we can take tenant from elsewhere, Like user
+    # session? They have it in session in Horizon, ehich correlates the teannt in keytsone token.
+    ext_management_system.with_provider_connection(:service     => "Network",
+                                                   :tenant_name => cloud_tenant.name) do |connection|
+      floating_ips.each do |floating_ip|
+        begin
+          connection.delete_floating_ip(floating_ip.ems_ref)
+        rescue StandardError => e
+          # The FloatingIp could have been deleted by another process
+          _log.info("Could not delete floating IP #{floating_ip} in EMS "\
+                    "'#{ext_management_system.name}'. Error: #{e}")
+        end
+        # Destroy it also in db, so we don't have to wait for refresh.
+        floating_ip.destroy
+      end
+    end
+  end
+
+  def destroy_if_failed
```

Comment:
This method seems odd to me.  Automate has access to the `raw_power_state` since it is a db column and the destroy method is also available to automate from the [remove_from_disk](https://github.com/ManageIQ/manageiq/blob/master/lib/miq_automation_engine/service_models/miq_ae_service_vm_or_template.rb#L159) method defined in the base class.

Is it really needed?